pkg/reconciler/inmemorychannel/inmemorychannel.go

```diff
@@ -0,0 +1,295 @@
+	dispatcherNamespace string,
+	dispatcherDeploymentName string,
+	dispatcherServiceName string,
+	inmemorychannelinformer messaginginformers.InMemoryChannelInformer,
+	deploymentInformer appsv1informers.DeploymentInformer,
+	serviceInformer corev1informers.ServiceInformer,
+	endpointsInformer corev1informers.EndpointsInformer,
+) *controller.Impl {
+
+	r := &Reconciler{
+		Base:                     reconciler.NewBase(opt, controllerAgentName),
+		dispatcherNamespace:      dispatcherNamespace,
+		dispatcherDeploymentName: dispatcherDeploymentName,
+		dispatcherServiceName:    dispatcherServiceName,
+		inmemorychannelLister:    inmemorychannelinformer.Lister(),
+		inmemorychannelInformer:  inmemorychannelinformer.Informer(),
+		deploymentLister:         deploymentInformer.Lister(),
+		serviceLister:            serviceInformer.Lister(),
+		endpointsLister:          endpointsInformer.Lister(),
+	}
+	r.impl = controller.NewImpl(r, r.Logger, ReconcilerName, reconciler.MustNewStatsReporter(ReconcilerName, r.Logger))
+
+	r.Logger.Info("Setting up event handlers")
+	inmemorychannelinformer.Informer().AddEventHandler(reconciler.Handler(r.impl.Enqueue))
+
+	// Set up watches for dispatcher resources we care about, since any changes to these
+	// resources will affect our Channels. So, set up a watch here, that will cause
+	// a global Resync for all the channels to take stock of their health when these change.
+	deploymentInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: FilterWithNameAndNamespace(dispatcherNamespace, dispatcherDeploymentName),
+		Handler:    r,
+	})
+	serviceInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: FilterWithNameAndNamespace(dispatcherNamespace, dispatcherServiceName),
+		Handler:    r,
+	})
+	endpointsInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: FilterWithNameAndNamespace(dispatcherNamespace, dispatcherServiceName),
+		Handler:    r,
+	})
+	return r.impl
+}
+
+// These 3 functions just cause a Global Resync of the channels, because any changes here
+// should be reflected onto the channels.
+func (r *Reconciler) OnAdd(obj interface{}) {
+	r.impl.GlobalResync(r.inmemorychannelInformer)
+}
+
+func (r *Reconciler) OnUpdate(old, new interface{}) {
+	r.impl.GlobalResync(r.inmemorychannelInformer)
+}
+
+func (r *Reconciler) OnDelete(obj interface{}) {
+	r.impl.GlobalResync(r.inmemorychannelInformer)
+}
+
+// Reconcile compares the actual state with the desired, and attempts to
+// converge the two. It then updates the Status block of the InMemoryChannel resource
+// with the current status of the resource.
+func (r *Reconciler) Reconcile(ctx context.Context, key string) error {
+	// Convert the namespace/name string into a distinct namespace and name.
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		logging.FromContext(ctx).Error("invalid resource key")
+		return nil
+	}
+
```

Comment:
Logging the key for all logging.FromContext will help. Something like `ctx := logging.With(ctx,  zap.Any("key", key)`

Reply:
knative/pkg controller does it for us automatigally.